### ISSUE TITLE
From hours to seconds: AI tools to detect animal calls

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11891]
From hours to seconds: AI tools to detect animal calls<br>
<br>
https://ift.tt/IqbRFef<br>
<br>
<br>
<br>

<div><div><div id="readability-page-1"><section><p>If you’re currently spending time listening to long audio recordings of birdcalls or animal noises, it’s likely that the last four years of advances in machine learning can save you a lot of time. I would like to help. I hope this blog post convinces you of two things: first, that near-state-of-the-art machine learning is easy to use even if you’re not a professional programmer, and second, that with a day or two of effort you can set up a pipeline that will reliably recognize any animal call (e.g. a birdcall), allowing you to record more data in more places. If you read this post and you want to try this, feel free to email me and I’ll pitch in to help.</p>
<p>The genesis of this idea came from my partner, who came back from a field naturalist talk having been told that it was currently impossible to build an automatic recogniser for the call of the Australian Powerful Owl, <em>Ninox strenua</em>. I felt that this couldn’t possibly be right in 2024, after the recent frenzy of investment in AI models and tooling. Indeed, it is possible! I hacked together <a href="https://ninoxstrenua.site/">ninoxstrenua.site</a>, which will take in an audio file (up to a couple of hundred mb) and tell you if it contains powerful owl calls:</p>
<p><span>
      <a href="https://www.seangoedecke.com/static/a226c4532b5e28c9bf23a83b14616d3d/1628f/owls.png" target="_blank" rel="noopener">
    <span></span>
  <img alt="owls" title="owls" src="https://www.seangoedecke.com/static/a226c4532b5e28c9bf23a83b14616d3d/fcda8/owls.png"></a>
    </span></p>
<p>I’m going to try to explain how to do this, assuming an audience that is not a professional programmer but has dabbled a bit with Python, and who has access to a reasonable volume of animal audio (say, a couple of hours).</p>
<p>Edit: Since writing this up, I learned about <a href="https://github.com/kahst/BirdNET-Analyzer">BirdNET</a>, which is an existing solution that will probably meet your needs. You’re probably better off using that instead of training your own model, unless you need it to be very fast or you’re willing to commit a lot of time to making a large enough dataset that can compete with BirdNET.</p>
<h2>Setting up your tools and accounts</h2>
<p>This is going to be the hardest part if you don’t do a lot of Python programming. Bear with it - it’s all downhill from here. First, you’re going to need to install <code>python3</code> on whatever system you have: <a href="https://www.python.org/downloads/">https://www.python.org/downloads/</a>. Next, you’re going to have to install a bunch of libraries for audio processing. In a macOS terminal, the command is something like <code>pip install pydub librosa soundfile datasets simpleaudio numpy huggingface_hub</code>. It might not work the first time. In the worst case, you’ll have to look up the error message and figure it out. The general tutorial for installing Python packages is <a href="https://packaging.python.org/en/latest/tutorials/installing-packages/">https://packaging.python.org/en/latest/tutorials/installing-packages/</a>.</p>
<p>As a general note, <a href="https://chat.openai.com/">ChatGPT</a> is pretty good at helping out with stuff like this: you might be able to paste in whatever error message you’re getting and have it figure it out.</p>
<p>You’ll have to sign up for a free account on <a href="https://huggingface.co/">https://huggingface.co/</a> - that’s the website that’s going to store your dataset and your model. I also strongly recommend signing up to <a href="https://cloud.lambdalabs.com/">https://cloud.lambdalabs.com/</a>. You’ll be using that website to rent out a <a href="https://en.wikipedia.org/wiki/Graphics_processing_unit">GPU</a> to train your model later on, which will cost a couple of bucks. If you really don’t want to do that, you can rent a GPU somewhere else, or rely on your own computer if it has a GPU already, but the setup will be harder.</p>
<h2>Chunking and labelling your dataset</h2>
<p>The next step is to gather up all your audio files. It doesn’t matter if you have one large file or a lot of small ones, but ideally you should have a couple of hundred instances of the animal call you’re looking for in there. Don’t use too much data, since you’re going to be processing it manually. You’ll want it all in <code>.wav</code> format. Start in a new directory, with all your files in a <code>/data</code> directory:</p>
<div><pre><code>workspace/
├─ data/
│  ├─ recording1.wav
│  ├─ moredata.wav</code></pre></div>
<p>This is going to form the training data for your model. Before you can train, you’ll need to do three things to your data: split it into even chunks, label each chunk (e.g. with “owl” and “not an owl”), and package it into a HuggingFace dataset. Now, from your workspace folder, run <a href="https://github.com/sgoedecke/birdcall-classifier-model/blob/master/chunk-raw-data.py">this script</a>. That will turn all your audio files into regular five-second chunks:</p>
<div><pre><code>workspace/
├─ segments/
│  ├─ recording1_segment_1.wav
│  ├─ recording1_segment_2.wav
│  ├─ recording1_segment_3.wav
│  ├─ moredata_segment_1.wav
├─ data/
│  ├─ recording1.wav
│  ├─ moredata.wav</code></pre></div>
<p>Now they’re ready to be labelled. You’ll need to split them up into two folders: <code>dataset/owls/</code> and <code>dataset/not-owls/</code>. This is the slow part. You could manually listen to all of these and move them into each folder - that would work fine. To make it quicker, I used <a href="https://github.com/sgoedecke/birdcall-classifier-model/blob/master/manual-audio-classifier.py">this script</a>. That will play them automatically, then prompt you to classify like this:</p>
<div><pre><code>sgoedecke@Seans-MacBook-Pro new-apr-24 % python3 /Users/sgoedecke/Code/birds/manual-audio-classifier.py
Processing sample 1 out of 22
Press ENTER to move to /not_owls, 'o' then ENTER for /owls, 'r' to replay: </code></pre></div>
<p>My advice here is to not worry about whether something is a “good enough” example of your animal call - even if it’s quiet, or cut off at the end or start of the segment, classify it as a “yes” and let the model sort it out.</p>
<p>I spent three hours doing this for my owl classifier. You could spend less, but the more time you spend here the better your model is going to be. Your model’s only as smart as the data you train it with. Eventually, you should have a structure like this:</p>
<div><pre><code>workspace/
├─ dataset/
├─ ├─ owls/
│  ├─ ├─ recording1_segment_1.wav
│  ├─ ├─ recording1_segment_3.wav
├─ ├─ not-owls/
│  ├─ ├─ recording1_segment_2.wav
│  ├─ ├─ moredata_segment_1.wav
├─ segments/
│  ├─ recording1_segment_1.wav
│  ├─ recording1_segment_2.wav
│  ├─ recording1_segment_3.wav
│  ├─ moredata_segment_1.wav
├─ data/
│  ├─ recording1.wav
│  ├─ moredata.wav</code></pre></div>
<p>Now it’s time to package this up into a dataset! <a href="https://github.com/sgoedecke/birdcall-classifier-model/blob/master/classified-audio-to-dataset.py">This script</a> will do it for you - you should edit the name of the dataset on line 42 to be whatever animal you’re doing. Call it anything you like. It’s not strictly necessary, but you should also probably edit the label names from <code>owl</code>/<code>not-owl</code> to whatever animal you’re using.</p>
<p>The script will prompt you for a <a href="https://huggingface.co/">HuggingFace</a> token at some point - go log in and generate one, making sure it has write permissions. If all goes well, you should be able to go to your HuggingFace profile and see your brand new dataset. After ten minutes or so, you should even be able to listen to some of the audio in it. Here’s what mine looks like:</p>
<p><span>
      <a href="https://www.seangoedecke.com/static/378d0bf383c52ae3f223830d7a833b7e/96191/hf_dataset.png" target="_blank" rel="noopener">
    <span></span>
  <img alt="hf_dataset" title="hf_dataset" src="https://www.seangoedecke.com/static/378d0bf383c52ae3f223830d7a833b7e/fcda8/hf_dataset.png"></a>
    </span></p>
<p>There should be at least several hundred entries in the dataset: if there aren’t, or if the audio doesn’t work, or the labelling isn’t right, then unfortunately you’ve gone wrong somewhere when running that last script.</p>
<h2>Training your model</h2>
<p>Now it’s time to train the model. We’re going to be using the SEW-D <a href="https://github.com/asappresearch/sew">model</a>, released in this September 2021 <a href="https://arxiv.org/abs/2109.06870">paper</a>. It’s a much smaller and faster version of Facebook’s wav2vec2 model, which was released about a year earlier. These details shouldn’t really matter to you - I mention them in case you’re curious and because I spent a long time trying out different options before I landed on SEW-D.</p>
<p>To train a model, you need GPUs. Otherwise we’ll be here for days. I recommend going to <a href="https://cloud.lambdalabs.com/">https://cloud.lambdalabs.com/</a> and launching an instance. The cheapest instance available is fine. If all goes well, you’ll only be using it for about ten minutes. Note that once you boot an instance, you’re paying for it until you terminate it, so don’t forget to terminate it when you’re done!</p>
<p>Copy the ssh command from <a href="https://cloud.lambdalabs.com/instances">https://cloud.lambdalabs.com/instances</a> and run it. After typing <code>yes</code> to the prompt, you should be in a terminal with a prompt like <code>ubuntu@146-235-202-184:~$</code>. Paste in these commands to make sure you’ll have everything you need:</p>
<div><pre><code>pip install soundfile librosa evaluate transformers pydub
pip install accelerate -U</code></pre></div>
<p>Now you can train your model by running <a href="https://github.com/sgoedecke/birdcall-classifier-model/blob/master/train-owl-model-sew-d.py">this script</a>. I typically do it by opening a python repl with <code>python3</code> and then just pasting the body of that file directly into the prompt. You’ll need to enter your HuggingFace token again, and update the dataset on line 18 from my dataset (<code>sgoedecke/powerful_owl_whatever</code> to your dataset, which will begin with your HuggingFace username and a forward slash). If you’re using different labels in your dataset, make sure you update those as well. Finally, update the name <a href="https://github.com/sgoedecke/birdcall-classifier-model/blob/master/train-owl-model-sew-d.py#L82">here</a> to be whatever you want to call your model. Now paste in the script!</p>
<p>This is going to take about 10-15 minutes. You’ll see some loading bars as the script downloads your model, and then a bunch of training output that looks something like this:</p>
<div><pre><code><span>&gt;&gt;</span><span>&gt;</span> trainer<span>.</span>train<span>(</span><span>)</span>
<span>{</span><span>'loss'</span><span>:</span> <span>0.6925</span><span>,</span> <span>'grad_norm'</span><span>:</span> <span>1.4354974031448364</span><span>,</span> <span>'learning_rate'</span><span>:</span> <span>6.521739130434783e-06</span><span>,</span> <span>'epoch'</span><span>:</span> <span>0.22</span><span>}</span>
<span>{</span><span>'loss'</span><span>:</span> <span>0.6624</span><span>,</span> <span>'grad_norm'</span><span>:</span> <span>3.0547149181365967</span><span>,</span> <span>'learning_rate'</span><span>:</span> <span>1.3043478260869566e-05</span><span>,</span> <span>'epoch'</span><span>:</span> <span>0.43</span><span>}</span>
<span>{</span><span>'loss'</span><span>:</span> <span>0.5909</span><span>,</span> <span>'grad_norm'</span><span>:</span> <span>7.781609535217285</span><span>,</span> <span>'learning_rate'</span><span>:</span> <span>1.956521739130435e-05</span><span>,</span> <span>'epoch'</span><span>:</span> <span>0.65</span><span>}</span>
<span>{</span><span>'loss'</span><span>:</span> <span>0.5691</span><span>,</span> <span>'grad_norm'</span><span>:</span> <span>6.032699108123779</span><span>,</span> <span>'learning_rate'</span><span>:</span> <span>2.608695652173913e-05</span><span>,</span> <span>'epoch'</span><span>:</span> <span>0.87</span><span>}</span>
<span>{</span><span>'eval_loss'</span><span>:</span> <span>0.4933532178401947</span><span>,</span> <span>'eval_precision'</span><span>:</span> <span>0.6876513317191283</span><span>,</span> <span>'eval_recall'</span><span>:</span> <span>0.9562289562289562</span><span>,</span> <span>'eval_f1'</span><span>:</span> <span>0.7999999999999999</span><span>,</span> <span>'eval_fbeta'</span><span>:</span> <span>0.728578758337609</span><span>,</span> <span>'eval_runtime'</span><span>:</span> <span>22.1506</span><span>,</span> <span>'eval_samples_per_second'</span><span>:</span> <span>28.532</span><span>,</span> <span>'eval_steps_per_second'</span><span>:</span> <span>0.903</span><span>,</span> <span>'epoch'</span><span>:</span> <span>1.0</span><span>}</span></code></pre></div>
<p>As this scrolls by, you should see the <code>loss</code> value decrease (it might tick up occasionally, but overall it should go down). The more important number is that <code>eval_f1</code> value that you’ll get at the end of each “epoch”, or block of training. We’re doing 15 epochs. At the end, <code>eval_f1</code> should be around <code>.94</code> (higher is better). This is the <a href="https://en.wikipedia.org/wiki/F-score">F1-score</a>, which represents a nice balance between the model’s “precision” (how many of the animal calls it notices) and its “recall” (the rate at which it makes accurate predictions). This is a pretty sensible way to measure how good a classification model is, so it’s what we’re using.</p>
<p>When the training is complete, your model should automatically upload itself to HuggingFace. You should be able to navigate to your profile on <a href="https://huggingface.co/">https://huggingface.co/</a> and see it under your models. If you see it, then you should be able to terminate your LambdaLabs instance.</p>
<h2>Using your model</h2>
<p>You’ve done all the hard work. Now you get to actually use your model on your own computer! Download <a href="https://github.com/sgoedecke/birdcall-classifier-model/blob/master/manual/infer.py">this script</a>, and update the model name <a href="https://github.com/sgoedecke/birdcall-classifier-model/blob/master/manual/infer.py#L34">here</a> to be the model you just uploaded in the last step. Now pick any of your audio files (ideally one you didn’t use to train the model, but it will work on those too), and run:</p>
<div><pre><code>python3 infer.py my-audio-file.wav</code></pre></div>
<p>After running for a few minutes, it should print some output like this (and also save it to a file):</p>
<div><pre><code>---------------------------------------------
Some of these are likely to be false-positives. Please listen to the actual audio segments to confirm.
Potential owls found:
File: ./data/WCS5_20231229_045900.wav
  Start: 00:13:20, End: 00:13:25, Chance: 0.75
  Start: 00:13:25, End: 00:13:30, Chance: 1.60
  Start: 00:13:30, End: 00:13:35, Chance: 3.44
  Start: 00:13:35, End: 00:13:40, Chance: 3.43</code></pre></div>
<p>And that’s it! Those represent where your model thought it recognized the animal call. The chance value is how confident the model was (here the first two might be false positives, but probably not the last two).</p>
<p>This is a lot of work for someone who doesn’t write a lot of Python for their day job (for what it’s worth, I write very little Python for my day job, and relied heavily on ChatGPT/Copilot to hack these scripts together). But once you do it, it stays done, and you can process an unlimited amount of audio very quickly. </p>
<p>A big thanks to my partner, and to James Deane from the <a href="https://warringal.org.au/">Warringal Conservation Society</a> for his advice and providing testing data. I think the potential of modern machine learning for audio classification is really under-utilized, and I’d love to see more people spend a few days setting this up instead of spending weeks listening to wildlife recordings. If you got stuck, and you’re in a position where you could benefit from a model like this, I am happy to help: email me at sean.goedecke@gmail.com.</p></section></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via www.seangoedecke.com https://ift.tt/XtiKa4W<br>
<br>
November 14, 2024 at 10:05AM
